pilopt/src/lib.rs

```diff
@@ -554,6 +562,128 @@ fn extract_constant_lookups<T: FieldElement>(pil_file: &mut Analyzed<T>) {
+/// TODO: the optimization is *NOT* applied to witness columns which are boolean constrained: intermediate polynomials get inlined in witness generation and
+/// the boolean constraint gets lost. Remove this limitation once intermediate polynomials are handled correctly in witness generation.
+fn replace_linear_witness_columns<T: FieldElement>(pil_file: &mut Analyzed<T>) {
+    let intermediate_definitions = pil_file.intermediate_definitions();
+
+    // We cannot remove arrays or array elements, so we filter them out.
+    // Also, we filter out columns that are used in public declarations.
+    // Also, we filter out columns that are boolean constrained.
+
+    let in_publics: HashSet<_> = pil_file
+        .public_declarations_in_source_order()
+        .map(|(_, pubd)| pubd.referenced_poly().name.clone())
+        .collect();
+
+    // pattern match identities looking for `w * (1 - w) = 0` and `(1 - w) * w = 0` constraints
+    let boolean_constrained_witnesses = pil_file
+        .identities
+        .iter()
+        .filter_map(|id| try_to_boolean_constrained(id))
+        .collect::<HashSet<_>>();
+
+    let keep = pil_file
+        .committed_polys_in_source_order()
+        .filter(|&(s, _)| {
+            s.is_array()
+                || in_publics.contains(&s.absolute_name)
+                || boolean_constrained_witnesses
+                    .intersection(&s.array_elements().map(|(_, poly_id)| poly_id).collect())
+                    .count()
+                    > 0
+        })
+        .map(|(s, _)| s.into())
+        .collect::<HashSet<PolyID>>();
+
+    let linear_polys = pil_file
+        .identities
+        .iter()
+        .enumerate()
+        .filter_map(|(index, id)| {
+            if let Identity::Polynomial(i) = id {
+                Some((index, i))
+            } else {
+                None
+            }
+        })
+        .filter_map(|(index, identity)| {
+            try_to_linearly_constrained(identity, &intermediate_definitions).map(|e| (index, e))
+        })
+        .filter(|(_, ((_, poly_id), _))| !keep.contains(poly_id))
+        .collect::<Vec<(usize, ((String, PolyID), _))>>();
+
+    let (identities_to_remove, intermediate_polys): (BTreeSet<_>, Vec<_>) = linear_polys
+        .iter()
+        .filter_map(|(index, ((name, old_poly_id), expression))| {
+            // Remove the definition. If this fails, we have already replaced it by an intermediate column.
+            pil_file.definitions.remove(name).map(|(symbol, value)| {
+                // sanity checks that we are looking at a single witness column
+                assert!(symbol.kind == SymbolKind::Poly(PolynomialType::Committed));
+                assert!(value.is_none());
+                assert!(symbol.length.is_none());
+
+                // we introduce a new intermediate symbol, so we need a new unique id
+                let id = pil_file
```

Comment:
This at least could be a function in Analyzed: return a new ID of a certain kind.

Reply:
Done.